### DISCUSSION TITLE
Realtime and RLS not working together when checking for user

### DISCUSSION THREAD

**Discussion by dustin-we** [supabase/supabase#11922] [Questions] [answered]
Hi,

I have a simple, blog-style app, which I use for testing.
I have a table "posts" that contains the posts the users write, users are referenced via "author_id" which is a foreign key to "auth_users".
My RLS policy looks like this:

![grafik](https://user-images.githubusercontent.com/44120716/214609667-b77668b4-bcca-4b05-bf22-0f08036ef1c4.png)

If I run a query to get all posts, I get the ones that are published, and the ones I wrote myself. Working as intended.

Now where things get weird. I subscribed to a realtimeChannel like this:

```
onMounted(() => {
  realtimeChannel = client.channel("public:posts").on(
    "postgres_changes",
    {
      event: "*",
      schema: "public",
      table: "posts",
    },
    (payload) => {
      console.log(payload);
      refreshPosts();
    }
  );

  realtimeChannel.subscribe();
});
```

But now I only get updates, inserts etc., if published=true on the DB. For me it looks like the channel subscription does not know about the user im logged in with? So it works for normal queries but not for the realtime subscription.

I am using Nuxt3 and the @nuxtjs/supabase package and followed its documentation:
https://supabase.nuxtjs.org/usage/composables/use-supabase-client

Thanks for your help!

**Comment by GaryAustin1** [comment]
Not sure about that repository, but with realtime-js you have to handle the jwt (supabase-js handles it for you).

In realtime-js there is a setAuth(jwt) call to apply a user jwt for RLS.

**Reply by dustin-we** [reply]
Thanks, although @nuxtjs/supabase already uses supabase-js under the hood. I also tried supabase-js as vanilla a few minutes ago, it also doesnt work.

**Comment by dustin-we** [comment]
Update: It also seems like if I delete something, even non-authenticated users, who, according to RLS, should not see the row, get notified that a row has been deleted.

So far, I have tested it with supabase-js, newest version, in a simple .js file, I also dont get the right updates.
Also, I remade the project, problem still persists.

**Reply by GaryAustin1** [reply]
Delete's always come thru, they ignore RLS.  This is why they only return id.  

This points to not having a user jwt when your subscription is running.

**Reply by dustin-we** [reply]
I might have found the issue. In my Nuxt code, I run the 'subscribe()' function in 'onMounted', I tried console logging the user at this point in time, it is null, so maybe it takes a while for the user to log in. Will report back!

**Reply by GaryAustin1** [reply]
The code you showed before you deleted it,  will not work.  You have to wait for something like getSession() to make sure the log in process from the redirect is finished.

**Reply by dustin-we** [reply]
Yeah, I am currently trying to figure out how to do that in Nuxt... Thank you!

**Reply by dustin-we** [reply]
Can you explain, how I would get this code to work? This is what I have currently:

```
const client = createClient(supabaseUrl, supabaseKey);

await client.auth.signInWithOAuth({
  provider: "azure",
  options: {
    scopes: "email",
  },
});

const session = await client.auth.getSession();

console.log(session);

const channel = client.channel("public:posts");

channel.on(
  "postgres_changes",
  {
    event: "*",
    schema: "public",
    table: "posts",
  },
  (payload) => console.log(payload)
);

channel.subscribe();

```

`console.log(session);`
returns no session

Thanks!

**Reply by GaryAustin1** [reply]
I would think your code would keep signing in each time, but I don't use signInWithOAuth.   When you call that you will get a new screen to log in, then you will be redirected to your app to either the siteURL or using the redirectTo URL (which you don't provide) and start running the code from that point again with the createClient code.

The app will start processing the redirect to log in the user, but you immediately call signInWithAuth again.

I would have getSession before that call at least.  And if session is null then do the signInWithOAuth, otherwise run my code with the returned user session from getSession.  (handling errors of course, but not needed for testing).

**Reply by dustin-we** [reply]
I am currently discussing this with supabase support. I will update this disussion as soon as we get it resolved. At the moment, it looks like Realtime does not use the JWT of the currently authenticated user in my app. 

```
onMounted(() => {
  realtimeChannel = client.channel("public:posts").on(
    "postgres_changes",
    {
      event: "*",
      schema: "public",
      table: "posts",
    },
    (payload) => {
      console.log(payload);
      refreshPosts();
    }
  );

  realtimeChannel.subscribe((status) => {
    if (status === "SUBSCRIBED") {
      client.realtime.setAuth(
      "My JWT",  
      );
    }
  });
});
```
If I set the realtime auth like this after subscribing, I get all updates.

**Comment by dustin-we** [comment, marked as answer]
I think I finally got it to work.

When using version 0.3.0 of @nuxtjs/supabase, neither useSupabaseAuthClient(), nor useSupabaseClient() could get me a session via .auth.getSession().

I downgraded to 0.2.6 of @nuxtjs/supabase, deleted cookies, changed code up a bit, since useSupabaseAuthClient() is not available in 0.2.6, and voila, it works like a charm.

I hope this helps someone struggling with the same problem. I will open a discussion on their repo, maybe I can even fix it myself, would be a great start to my open source journey.

Thanks for your patience, also thanks to supabase support :)
